sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala

```diff
@@ -410,6 +410,20 @@ case class PrettyAttribute(
   override def nullable: Boolean = true
 }
 
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(child: Expression, scope: AttributeSeq)
```

Comment:
shall we put `Seq[Attribute]` and include it as children?

Reply:
> shall we put Seq[Attribute]

yes, done:
```diff
-case class ScopedExpression(child: Expression, scope: AttributeSeq)
+case class ScopedExpression(child: Expression, scope: Seq[Attribute])
```

> include it as children?

as in
```diff
 case class ScopedExpression(child: Expression, scope: Seq[Attribute])
-  extends UnaryExpression with Unevaluable {
+  extends Expression with Unevaluable {
   override def dataType: DataType = child.dataType
   override def nullable: Boolean = child.nullable
   override def prettyName: String = "scoped"
 
   override def sql: String = s"$prettyName(${child.sql}, $scope)"
 
-  override def withNewChildInternal(newChild: Expression): ScopedExpression = copy(child = newChild)
+  override def withNewChildren(newChildren: Seq[Expression]): ScopedExpression = {
+    val scope = newChildren.tail
+    assert(scope.forall(_.isInstanceOf[Attribute]), "Scope children have to be attributes")
+    copy(child = newChildren.head, scope = scope.map(_.asInstanceOf[Attribute]))
+  }
 }
```